builtin/providers/aws/cloudfront_distribution_configuration_structure.go

```diff
@@ -0,0 +1,942 @@
+// complex structures which are necessary to properly build the
+// DistributionConfig structure.
+//
+// Used by the aws_cloudfront_distribution Create and Update functions.
+func expandDistributionConfig(d *schema.ResourceData) *cloudfront.DistributionConfig {
+	distributionConfig := &cloudfront.DistributionConfig{
+		CacheBehaviors:       expandCacheBehaviors(d.Get("cache_behavior").(*schema.Set)),
+		CustomErrorResponses: expandCustomErrorResponses(d.Get("custom_error_response").(*schema.Set)),
+		DefaultCacheBehavior: expandDefaultCacheBehavior(d.Get("default_cache_behavior").(*schema.Set).List()[0].(map[string]interface{})),
+		Enabled:              aws.Bool(d.Get("enabled").(bool)),
+		Origins:              expandOrigins(d.Get("origin").(*schema.Set)),
+		PriceClass:           aws.String(d.Get("price_class").(string)),
+	}
+	// This sets CallerReference if it's still pending computation (ie: new resource)
+	if v, ok := d.GetOk("caller_reference"); ok == false {
+		distributionConfig.CallerReference = aws.String(time.Now().Format(time.RFC3339Nano))
+	} else {
+		distributionConfig.CallerReference = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("comment"); ok {
+		distributionConfig.Comment = aws.String(v.(string))
+	} else {
+		distributionConfig.Comment = aws.String("")
+	}
+	if v, ok := d.GetOk("default_root_object"); ok {
+		distributionConfig.DefaultRootObject = aws.String(v.(string))
+	} else {
+		distributionConfig.DefaultRootObject = aws.String("")
+	}
+	if v, ok := d.GetOk("logging_config"); ok {
+		distributionConfig.Logging = expandLoggingConfig(v.(*schema.Set).List()[0].(map[string]interface{}))
+	} else {
+		distributionConfig.Logging = expandLoggingConfig(nil)
+	}
+	if v, ok := d.GetOk("aliases"); ok {
+		distributionConfig.Aliases = expandAliases(v.(*schema.Set))
+	} else {
+		distributionConfig.Aliases = expandAliases(schema.NewSet(aliasesHash, []interface{}{}))
+	}
+	if v, ok := d.GetOk("restrictions"); ok {
+		distributionConfig.Restrictions = expandRestrictions(v.(*schema.Set).List()[0].(map[string]interface{}))
+	}
+	if v, ok := d.GetOk("viewer_certificate"); ok {
+		distributionConfig.ViewerCertificate = expandViewerCertificate(v.(*schema.Set).List()[0].(map[string]interface{}))
+	}
+	if v, ok := d.GetOk("web_acl_id"); ok {
+		distributionConfig.WebACLId = aws.String(v.(string))
+	} else {
+		distributionConfig.WebACLId = aws.String("")
+	}
+	return distributionConfig
+}
+
+// Unpack the *cloudfront.DistributionConfig variable and set resource data.
+// Calls out to flatten functions to convert the DistributionConfig
+// sub-structures to their respective attributes in the
+// aws_cloudfront_distribution resource.
+//
+// Used by the aws_cloudfront_distribution Read function.
+func flattenDistributionConfig(d *schema.ResourceData, distributionConfig *cloudfront.DistributionConfig) {
+	d.Set("origins", flattenOrigins(distributionConfig.Origins))
```

Comment:
for anything that is not a simple value like `int` or `string`, we need to check the error from `d.Set` to ensure it was set correctly